src/v2/style-guide/index.md

```diff
@@ -0,0 +1,1802 @@
+
+- It generally takes more time to navigate through nested sub-directories, than scrolling through a single `components` directory.
+- Name conflicts (e.g. multiple `ButtonDelete.vue` components) make it more difficult to quickly navigate to a specific component in a code editor.
+- Refactoring becomes more difficult, because find-and-replace often isn't sufficient to update relative references to a moved component.
+
+{% raw %}</details>{% endraw %}
+
+{% raw %}<div class="style-example example-bad">{% endraw %}
+#### Bad
+
+```
+components/
+|- ClearSearchButton.vue
+|- ExcludeFromSearchInput.vue
+|- LaunchOnStartupCheckbox.vue
+|- RunSearchButton.vue
+|- SearchInput.vue
+|- TermsCheckbox.vue
+```
+{% raw %}</div>{% endraw %}
+
+{% raw %}<div class="style-example example-good">{% endraw %}
+#### Good
+
+```
+components/
+|- SearchButtonClear.vue
+|- SearchButtonRun.vue
+|- SearchInputQuery.vue
+|- SearchInputExcludeGlob.vue
+|- SettingsCheckboxTerms.vue
+|- SettingsCheckboxLaunchOnStartup.vue
+```
+{% raw %}</div>{% endraw %}
+
+
+
+### Self-closing components <sup data-p="b">strongly recommended</sup>
+
+**Components with no content should be self-closing in [single-file components](single-file-components.html), string templates, and [JSX](render-function.html#JSX) - but never in DOM templates.**
+
+Components that self-close communicate that they not only have no content, but are **meant** to have no content. It's the difference between a blank page in a book and one labeled "This page intentionally left blank." Your code is also cleaner without the unnecessary closing tag.
+
+Unfortunately, HTML doesn't allow custom elements to be self-closing - only [official "void" elements](https://www.w3.org/TR/html/syntax.html#void-elements). That's why the strategy is only possible when Vue's template compiler can reach the template before the DOM, then serve the DOM spec-compliant HTML.
+
+{% raw %}<div class="style-example example-bad">{% endraw %}
+#### Bad
+
+``` html
+<!-- In single-file components, string templates, and JSX -->
+<MyComponent></MyComponent>
+```
+
+``` html
+<!-- In DOM templates -->
+<my-component/>
+```
+{% raw %}</div>{% endraw %}
+
+{% raw %}<div class="style-example example-good">{% endraw %}
+#### Good
+
+``` html
+<!-- In single-file components, string templates, and JSX -->
+<MyComponent/>
+```
+
+``` html
+<!-- In DOM templates -->
+<my-component></my-component>
+```
+{% raw %}</div>{% endraw %}
+
+
+
+### Component name casing in templates <sup data-p="b">strongly recommended</sup>
+
+**Component names should always be PascalCase in [single-file components](single-file-components.html) and string templates - but kebab-case in DOM templates.**
+
+PascalCase in templates improves readability, because it's easier to tell at a glance what's an element and what's a component. Unfortunately, PascalCase is invalid for a custom element in HTML, so a DOM template must still use kebab-case.
```

Comment:
I think using kebab-case for component names is also quite clear to me (HTML tag names don't contain `-`) and we can always use kebab-case in template strings/SFCs/DOM templates, which might be even more consistent IMO. PascalCase component name smells a little bit like JSX... What about “kebab-case in templates, PascalCase in JSX”？

Reply:
The PascalCase provides an extra layer of meaning in your app, by differentiating components from elements visually. Additionally, having the difference in convention between the two contexts (in-DOM components vs others) carries even more information, acting as a warning that you're in the DOM, so some limitations apply.

Using kekab-case everywhere doesn't have similar advantages, so "smells like JSX" isn't a good enough reason to reject a convention with clear benefits. 🙂